nixos/libvirtd: add dbus support

andre4ik3 opened the pull request:
Adds a simple module and NixOS tests for libvirt-dbus.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [x] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

andre4ik3 commented:
CC @Sigmanificient who was asking about libvirt-dbus testing in https://github.com/NixOS/nixpkgs/pull/447197.

maevii reviewed (changes requested):
Diff looks good, but I feel like this should be added to [the libvirtd module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/libvirtd.nix) directly instead of being separate like this, especially since it just extends it with a `dbus` option. Is there any reason I'm missing why it shouldn't?

andre4ik3 commented:
Rebased and fixed conflicts, and moved to the main `libvirtd.nix` file. (Originally I wanted to put it as a separate file to avoid cluttering the main module)

azey7f commented on nixos/doc/manual/release-notes/rl-2605.section.md:
```diff
@@ -22,4 +22,4 @@
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->
 
-- Create the first release note entry in this section!
+- Libvirt gains support for D-Bus via libvirt-dbus. Available as [virtualisation.libvirtd.dbus.enable](#opt-virtualisation.libvirtd.dbus.enable).
```
```suggestion
- `virtualisation.libvirtd` gained support for D-Bus via libvirt-dbus. Available as [virtualisation.libvirtd.dbus.enable](#opt-virtualisation.libvirtd.dbus.enable).
```

I'm also not sure whether this is significant enough to warrant a release note, but I guess it doesn't hurt. If anything, I'd consider this eligible to backport since it only adds functionality.

andre4ik3 replied:
Removed changelog entry.

maevii reviewed (commented):
Few minor things, otherwise it looks good. Both the `libvirtd` and the new `libvirt-dbus` tests also pass locally.

About the test though, I feel like it could instead be a subtest in `libvirtd` as well - both for making sure that `libvirtd` updates don't break `libvirt-dbus`, and vice versa.

andre4ik3 committed:
nixos/libvirtd: add dbus support

andre4ik3 commented on nixos/modules/virtualisation/libvirtd.nix:
```diff
@@ -449,7 +467,7 @@ in
     services.firewalld.packages = [ cfg.package ];
 
-    systemd.packages = [ cfg.package ];
+    systemd.packages = [ cfg.package ] ++ lib.optional cfg.dbus.enable cfg.dbus.package;
```
Libvirt comes with its own systemd service file, that needs to go into `systemd.packages` so it gets picked up by systemd, no? The D-Bus service activates the systemd service.

azey7f replied:
Sorry, you're right! I missed this was `systemd.packages` and not `environment.systemPackages`. No problem then.

andre4ik3 commented on pkgs/by-name/li/libvirt-dbus/package.nix:
```diff
@@ -60,8 +61,13 @@ stdenv.mkDerivation (finalAttrs: {
+      rev-prefix = "v";
+    };
+    tests = {
+      inherit (nixosTests) libvirt-dbus;
```
Oh I changed it but forgot to add it before I amended the commit 😅

maevii reviewed (approved):
Test runs fine and the diff LGTM.

cc `libvirtd` package maintainers since the module doesn't have any listed:
@fpletz @globin @lovesegfault

Aleksanaa commented on nixos/modules/virtualisation/libvirtd.nix:
```diff
@@ -426,15 +431,28 @@ in
+        };
+      }
+      (lib.mkIf cfg.dbus.enable {
+        users.libvirtdbus = {
```
This user name is hardcoded in `share/dbus-1/system.d/org.libvirt.conf` and `share/polkit-1/rules.d/libvirt-dbus.rules`.
